fft/src/KokkosFFT_FFTW_Types.hpp

```diff
@@ -0,0 +1,158 @@
+template <typename ExecutionSpace, typename T1, typename T2>
+struct ScopedFFTWPlan {
+ private:
+  using floating_point_type = KokkosFFT::Impl::base_floating_point_type<T1>;
+  using plan_type =
+      std::conditional_t<std::is_same_v<floating_point_type, float>, fftwf_plan,
+                         fftw_plan>;
+  plan_type m_plan;
+
+ public:
+  template <typename InScalarType, typename OutScalarType>
+  ScopedFFTWPlan(const ExecutionSpace &exec_space, int rank, const int *n,
+                 int howmany, InScalarType *in, const int *inembed, int istride,
+                 int idist, OutScalarType *out, const int *onembed, int ostride,
+                 int odist, [[maybe_unused]] int sign, unsigned flags) {
+    init_threads(exec_space);
+    constexpr auto type = fftw_transform_type<T1, T2>::type();
+    if constexpr (type == KokkosFFT::Impl::FFTWTransformType::R2C) {
+      m_plan =
+          fftwf_plan_many_dft_r2c(rank, n, howmany, in, inembed, istride, idist,
+                                  out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::D2Z) {
+      m_plan =
+          fftw_plan_many_dft_r2c(rank, n, howmany, in, inembed, istride, idist,
+                                 out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::C2R) {
+      m_plan =
+          fftwf_plan_many_dft_c2r(rank, n, howmany, in, inembed, istride, idist,
+                                  out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::Z2D) {
+      m_plan =
+          fftw_plan_many_dft_c2r(rank, n, howmany, in, inembed, istride, idist,
+                                 out, onembed, ostride, odist, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::C2C) {
+      m_plan =
+          fftwf_plan_many_dft(rank, n, howmany, in, inembed, istride, idist,
+                              out, onembed, ostride, odist, sign, flags);
+    } else if constexpr (type == KokkosFFT::Impl::FFTWTransformType::Z2Z) {
+      m_plan = fftw_plan_many_dft(rank, n, howmany, in, inembed, istride, idist,
+                                  out, onembed, ostride, odist, sign, flags);
+    }
+  }
+
+  ~ScopedFFTWPlan() noexcept {
+    cleanup_threads();
+    if constexpr (std::is_same_v<plan_type, fftwf_plan>) {
+      fftwf_destroy_plan(m_plan);
+    } else {
+      fftw_destroy_plan(m_plan);
+    }
+  }
+
+  ScopedFFTWPlan()                                  = delete;
+  ScopedFFTWPlan(const ScopedFFTWPlan &)            = delete;
+  ScopedFFTWPlan &operator=(const ScopedFFTWPlan &) = delete;
+  ScopedFFTWPlan &operator=(ScopedFFTWPlan &&)      = delete;
+  ScopedFFTWPlan(ScopedFFTWPlan &&)                 = delete;
+
+  plan_type plan() const noexcept { return m_plan; }
+
+ private:
+  void init_threads([[maybe_unused]] const ExecutionSpace &exec_space) {
```

Comment:
```suggestion
  static void init_threads([[maybe_unused]] const ExecutionSpace &exec_space) noexcept {
```

The function is independent of the state of the object.